### PR TITLE
lambda_event - add support for maximum_batching_window_in_seconds

### PR DESCRIPTION
**Pull request by alinabuzachis**
##### SUMMARY

Closes https://github.com/ansible-collections/amazon.aws/issues/1995
Fix when batch_size is greater than 10, by enabling support for setting `maximum_batching_window_in_seconds`.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

lambda_event

### CODE REVIEW

File: plugins/modules/lambda_event.py [hunk 1]
```diff
@@ -252,13 +266,32 @@ def lambda_event_stream(module, client):
...
+            else:
+                if source_params["source_arn"].endswith(".fifo"):
+                    if source_params["batch_size"] > 10:
+                        module.fail_json(msg="For FIFO queues the maximum batch_size is 10.")
+                    if source_params.get("maximum_batching_window_in_seconds"):
+                        module.fail_json(
+                            msg="maximum_batching_window_in_seconds is not supported by Amazon SQS FIFO event sources."
+                        )
+                else:
+                    if source_params["batch_size"] > 10000 or source_params["batch_size"] < 100:
```
Comment (mandar242) [inline plugins/modules/lambda_event.py:283]:
```suggestion
                    if not (100 <= source_params["batch_size"] <= 10000):
```

File: plugins/modules/lambda_event.py [hunk 2]
```diff
@@ -252,13 +266,32 @@ def lambda_event_stream(module, client):
...
+                        module.fail_json(msg="For standard queue batch_size must be between 100 and 10000.")
+
+        elif module.params["event_source"].lower() == "stream":
+            if not source_params.get("batch_size"):
+                # Default 100 for streams.
+                source_params["batch_size"] = 100
+            elif source_params["batch_size"] > 10000 or source_params["batch_size"] < 100:
```
Comment (mandar242) [inline plugins/modules/lambda_event.py:290]:
```suggestion
            elif not (100 <= source_params["batch_size"] <= 10000):
```

**Review by mandar242 — APPROVED** [review]
Overall LGTM!

File: plugins/modules/lambda_event.py [hunk 3]
```diff
@@ -84,6 +90,14 @@
         elements: str
         choices: [ReportBatchItemFailures]
         version_added: 5.5.0
+      maximum_batching_window_in_seconds:
+        description:
+          - The maximum amount of time, in seconds, that Lambda spends gathering records before invoking the function.
+          - You can configure I(maximum_batching_window_in_seconds) to any value from C(0) seconds to C(300) seconds in increments of seconds.
+          - For streams and Amazon SQS event sources, when I(batch_size) to a value greater than C(10), I(maximum_batching_window_in_seconds) defaults to C(1).
```
Comment (abikouo) [inline plugins/modules/lambda_event.py:97]:
```suggestion
          - For streams and Amazon SQS event sources, when I(batch_size) is set to a value greater than C(10), I(maximum_batching_window_in_seconds) defaults to C(1).
```

File: plugins/modules/lambda_event.py [hunk 4]
```diff
@@ -252,13 +266,32 @@ def lambda_event_stream(module, client):
...
+
+        elif module.params["event_source"].lower() == "stream":
+            if not source_params.get("batch_size"):
+                # Default 100 for streams.
+                source_params["batch_size"] = 100
+            elif source_params["batch_size"] > 10000 or source_params["batch_size"] < 100:
+                module.fail_json(msg="batch_size for streams must be between 100 and 10000")
+
+        if source_params["batch_size"] > 10 and not source_params.get("maximum_batching_window_in_seconds"):
+            source_params["maximum_batching_window_in_seconds"] = 1
```
Comment (abikouo) [inline plugins/modules/lambda_event.py:294]: The parameter description says 
``For streams and Amazon SQS event sources, when I(batch_size) is set to a value greater than C(10), I(maximum_batching_window_in_seconds) defaults to C(1).``
therefore this code isn't supposed to move at line ``271`` (to be applied for SQS queue only)?

**Review by abikouo — CHANGES_REQUESTED** [review]

File: plugins/modules/lambda_event.py (hunk 4 above)
Comment (alinabuzachis) [reply plugins/modules/lambda_event.py:294]: Shouldn't it be applied `For streams and Amazon SQS event sources`?

Comment (abikouo) [reply plugins/modules/lambda_event.py:294]: if it applies for these 2 types only this means somewhere we need to run a condition before doing this

Comment (alinabuzachis) [reply plugins/modules/lambda_event.py:294]: The module supports only two types of `event_sources` `choices: ['stream', 'sqs']` in any case.

**Review by abikouo — APPROVED** [review]

File: plugins/modules/lambda_event.py [hunk 5]
```diff
@@ -252,13 +267,32 @@ def lambda_event_stream(module, client):
...
+                if source_params["source_arn"].endswith(".fifo"):
+                    if source_params["batch_size"] > 10:
+                        module.fail_json(msg="For FIFO queues the maximum batch_size is 10.")
+                    if source_params.get("maximum_batching_window_in_seconds"):
+                        module.fail_json(
+                            msg="maximum_batching_window_in_seconds is not supported by Amazon SQS FIFO event sources."
+                        )
+                else:
+                    if not (100 <= source_params["batch_size"] <= 10000):
+                        module.fail_json(msg="For standard queue batch_size must be between 100 and 10000.")
```
Comment (tremble) [inline plugins/modules/lambda_event.py:285]: I'm not sure I like the depth of these nested if-statements this feels like the kind of tangled nest of logic that can come back to bite us.

**Review by tremble — COMMENTED** [review]
I'm trying to untangle the logic here.

Is this having the practical effect of changing the defaults?  If so we shouldn't be backporting it.

File: plugins/modules/lambda_event.py (hunk 5 above)
Comment (alinabuzachis) [reply plugins/modules/lambda_event.py]: I tried to implement the checks for batch_size and maximum_batching_window_in_seconds as explained in the documentation https://boto3.amazonaws.com/v1/documentation/api/1.26.78/reference/services/lambda.html#Lambda.Client.create_event_source_mapping. However, I also cleaned a bit now.

What the documentation was reporting as default value for batch_size was wrong. It shouldn't always default to 100. I fixed that and I also added the breaking_change section in the changelog. backport-7 label is also gone.

File: plugins/modules/lambda_event.py [hunk 6]
```diff
@@ -54,22 +54,28 @@
...
-          -  The largest number of records that AWS Lambda will retrieve from your event source at the time of invoking your function.
-        default: 100
+          - The largest number of records that AWS Lambda will retrieve from your event source at the time of invoking your function.
+          - Amazon Kinesis - Default C(100). Max C(10000).
+          - Amazon DynamoDB Streams - Default C(100). Max C(10000).
+          - Amazon Simple Queue Service - Default C(10). For standard queues the max is C(10000). For FIFO queues the max is C(10).
+          - Amazon Managed Streaming for Apache Kafka - Default C(100). Max C(10000).
+          - Self-managed Apache Kafka - Default C(100). Max C(10000).
+          - Amazon MQ (ActiveMQ and RabbitMQ) - Default C(100). Max C(10000).
+          - DocumentDB - Default C(100). Max C(10000).
```
Comment (hakbailey) [inline plugins/modules/lambda_event.py:78]:
```suggestion
          - Amazon Kinesis - Default V(100). Max V(10000).
          - Amazon DynamoDB Streams - Default V(100). Max V(10000).
          - Amazon Simple Queue Service - Default V(10). For standard queues the max is V(10000). For FIFO queues the max is V(10).
          - Amazon Managed Streaming for Apache Kafka - Default V(100). Max V(10000).
          - Self-managed Apache Kafka - Default V(100). Max V(10000).
          - Amazon MQ (ActiveMQ and RabbitMQ) - Default V(100). Max V(10000).
          - DocumentDB - Default V(100). Max V(10000).
```

File: plugins/modules/lambda_event.py [hunk 7]
```diff
@@ -84,6 +90,15 @@
         elements: str
         choices: [ReportBatchItemFailures]
         version_added: 5.5.0
+      maximum_batching_window_in_seconds:
+        description:
+          - The maximum amount of time, in seconds, that Lambda spends gathering records before invoking the function.
+          - You can configure I(maximum_batching_window_in_seconds) to any value from C(0) seconds to C(300) seconds in increments of seconds.
+          - For streams and Amazon SQS event sources, when I(batch_size) is set to a value greater than C(10), I(maximum_batching_window_in_seconds)
+            defaults to C(1).
```
Comment (hakbailey) [inline plugins/modules/lambda_event.py:98]:
```suggestion
          - You can configure O(maximum_batching_window_in_seconds) to any value from V(0) seconds to V(300) seconds in increments of seconds.
          - For streams and Amazon SQS event sources, when O(batch_size) is set to a value greater than V(10), O(maximum_batching_window_in_seconds)
            defaults to V(1).
```

File: plugins/modules/lambda_event.py [hunk 8]
```diff
@@ -252,13 +267,31 @@ def lambda_event_stream(module, client):
...
-    batch_size = source_params.get("batch_size")
-    if batch_size:
-        try:
-            source_params["batch_size"] = int(batch_size)
-        except ValueError:
-            module.fail_json(msg=f"Source parameter 'batch_size' must be an integer, found: {batch_size}")
+    if state == "present":
+        if module.params["event_source"].lower() == "sqs":
+            # Default 10. For standard queues the max is 10,000. For FIFO queues the max is 10.
+            source_params.setdefault("batch_size", 10)
```
Comment (hakbailey) [inline plugins/modules/lambda_event.py:273]: Do we need to set these default values or can we just not send them and let AWS set the defaults?

File: plugins/modules/lambda_event.py [hunk 9]
```diff
@@ -252,13 +267,31 @@ def lambda_event_stream(module, client):
...
+
+            if source_params["source_arn"].endswith(".fifo"):
+                if source_params["batch_size"] > 10:
+                    module.fail_json(msg="For FIFO queues the maximum batch_size is 10.")
```
Comment (hakbailey) [inline plugins/modules/lambda_event.py:277]: Would it make more sense to move these validation checks to the `validate_params` function?

File: plugins/modules/lambda_event.py (hunk 5 above)
Comment (tremble) [reply plugins/modules/lambda_event.py]: > I tried to implement the checks for batch_size and maximum_batching_window_in_seconds as explained in the documentation

If it needs to be complex, then it needs to be complex.  I was thinking more that it probably needs to be pulled out into its own function that we can unit test: I don't think we're likely to create integration tests that go through the various combinations that are possible here...

Comment (alinabuzachis) [reply plugins/modules/lambda_event.py]: I added unit tests instead. Thanks.

File: plugins/modules/lambda_event.py (hunk 8 above)
Comment (alinabuzachis) [reply plugins/modules/lambda_event.py]: We need to handle them.

File: plugins/modules/lambda_event.py (hunk 9 above)
Comment (alinabuzachis) [reply plugins/modules/lambda_event.py]: I created a different function.

File: plugins/modules/lambda_event.py (hunk 6 above)
Comment (tremble) [reply plugins/modules/lambda_event.py]: Where's `V()` come from?  last I saw it was supposed to be `C()` for values...

File: plugins/modules/lambda_event.py [hunk 10]
```diff
@@ -228,6 +244,37 @@ def get_qualifier(module):
...
 
+def set_default_values(module, source_params):
+    _source_params_cpy = copy.deepcopy(source_params)
+
+    if module.params["event_source"].lower() == "sqs":
+        # Default 10. For standard queues the max is 10,000. For FIFO queues the max is 10.
+        _source_params_cpy.setdefault("batch_size", 10)
+
+        if source_params["source_arn"].endswith(".fifo"):
+            if source_params["batch_size"] > 10:
```
Comment (hakbailey) [inline plugins/modules/lambda_event.py:255]: I don't think `batch_size` is required, so this should either read from `_source_params_cpy["batch_size"]` which has the default set or it should be `source_params.get("batch_size", 10)` so this doesn't raise an error.

File: plugins/modules/lambda_event.py [hunk 11]
```diff
@@ -228,6 +244,37 @@ def get_qualifier(module):
...
+                module.fail_json(msg="For FIFO queues the maximum batch_size is 10.")
+            if source_params.get("maximum_batching_window_in_seconds"):
+                module.fail_json(
+                    msg="maximum_batching_window_in_seconds is not supported by Amazon SQS FIFO event sources."
+                )
+        else:
+            if not (100 <= source_params["batch_size"] <= 10000):
```
Comment (hakbailey) [inline plugins/modules/lambda_event.py:262]: Same here as above.

File: tests/unit/plugins/modules/test_lambda_event.py [hunk 12]
```diff
@@ -375,3 +384,143 @@ def test_lambda_event_stream_update_event(
...
         assert dict(changed=True, events=event_source_updated) == result
         client.update_event_source_mapping.assert_called_once_with(**api_params)
+
+
+@pytest.mark.parametrize(
+    "params, expected, exception, message, source_type",
+    [
+        (
+            {
+                "source_arn": "arn:aws:sqs:us-east-1:123456789012:ansible-test-28277052.fifo",
```
Comment (hakbailey) [inline tests/unit/plugins/modules/test_lambda_event.py:394]: We should probably also include a unit test for the `sqs` source type that does not include passed `batch_size` and `maximum_batching_window_in_seconds` to ensure default values are set as expected.

**Review by hakbailey — CHANGES_REQUESTED** [review]

File: plugins/modules/lambda_event.py (hunk 10 above)
Comment (alinabuzachis) [reply plugins/modules/lambda_event.py]: I used _source_params_cpy instead.

File: tests/unit/plugins/modules/test_lambda_event.py (hunk 12 above)
Comment (alinabuzachis) [reply tests/unit/plugins/modules/test_lambda_event.py:394]: Done!

**Comment by hakbailey** [comment]
recheck

**Review by hakbailey — APPROVED** [review]

**Pull request merged**